miniapps/shifted/extrapolate.cpp

```diff
@@ -0,0 +1,219 @@
+//     mpirun -np 4 extrapolate -rs 5 -p 1 -eo 2
+//     mpirun -np 4 extrapolate -rs 5 -p 1 -et 1 -eo 1 -dg 1
+//     mpirun -np 4 extrapolate -m "../../data/inline-hex.mesh" -eo 1 -rs 1
+//     mpirun -np 4 extrapolate -m "../../data/inline-hex.mesh" -p 1 -eo 1 -rs 1
+
+#include "extrapolator.hpp"
+
+using namespace std;
+using namespace mfem;
+
+int problem = 0;
+
+double domainLS(const Vector &coord)
+{
+   // Map from [0,1] to [-1,1].
+   const int dim = coord.Size();
+   const double x = coord(0)*2.0 - 1.0,
+                y = (dim > 1) ? coord(1)*2.0 - 1.0 : 0.0,
+                z = (dim > 2) ? coord(2)*2.0 - 1.0 : 0.0;
+
+   switch (problem)
+   {
+      case 0:
+      {
+         // Sphere.
+         return 0.75 - sqrt(x*x + y*y + z*z + 1e-12);
+      }
+      case 1:
+      {
+         // Star.
+         MFEM_VERIFY(dim > 1, "Problem 1 is not applicable to 1D.");
+
+         return 0.60 - sqrt(x*x + y*y + z*z + 1e-12) +
+                0.25 * (y*y*y*y*y + 5.0*x*x*x*x*y - 10.0*x*x*y*y*y) /
+                pow(x*x + y*y + z*z + 1e-12, 2.5) *
+                std::cos(0.5*M_PI * z / 0.6);
+      }
+      default: MFEM_ABORT("Bad option for --problem!"); return 0.0;
+   }
+}
+
+double solution0(const Vector &coord)
+{
+   // Map from [0,1] to [-1,1].
+   const int dim = coord.Size();
+   const double x = coord(0)*2.0 - 1.0 + 0.25,
+                y = (dim > 1) ? coord(1)*2.0 - 1.0 : 0.0,
+                z = (dim > 2) ? coord(2)*2.0 - 1.0 : 0.0;
+
+   return std::cos(M_PI * x) * std::cos(M_PI * y) * std::cos(M_PI * z);
+}
+
+void PrintNorm(int myid, Vector &v, std::string text)
+{
+   double norm = v.Norml1();
+   MPI_Allreduce(MPI_IN_PLACE, &norm, 1, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
+   if (myid == 0)
+   {
+      std::cout << std::setprecision(12) << std::fixed
+                << text << norm << std::endl;
+   }
+}
+
+void PrintIntegral(int myid, ParGridFunction &g, std::string text)
+{
+   ConstantCoefficient zero(0.0);
+   double norm = g.ComputeL1Error(zero);
+   if (myid == 0)
+   {
+      std::cout << std::setprecision(12) << std::fixed
+                << text << norm << std::endl;
+   }
+}
+
+int main(int argc, char *argv[])
+{
+   // Initialize MPI.
+   MPI_Session mpi;
+   int myid = mpi.WorldRank();
+
+   // Parse command-line options.
+   const char *mesh_file = "../../data/inline-quad.mesh";
+   int rs_levels = 2;
+   Extrapolator::XtrapType ex_type   = Extrapolator::ASLAM;
+   AdvectionOper::AdvectionMode dg_mode = AdvectionOper::HO;
+   int ex_order = 1;
+   int order = 2;
+   double distance = 0.35;
+   bool vis_on = true;
+   int vis_steps_cnt = 50;
+
+   OptionsParser args(argc, argv);
+   args.AddOption(&mesh_file, "-m", "--mesh",
+                  "Mesh file to use.");
+   args.AddOption(&rs_levels, "-rs", "--refine-serial",
+                  "Number of times to refine the mesh uniformly in serial.");
+   args.AddOption((int*)&ex_type, "-et", "--extrap-type",
```

Comment:
We should check wether the inputs for `ex_type` and `dg_mode` are valid.

Reply:
Done.